x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -67,3 +74,171 @@ func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types
+	}
+
+	consAddr, err := validator.GetConsAddr()
+	if err != nil {
+		return sdk.ValAddress{}, tmprotocrypto.PublicKey{}, err
+	}
+
+	pubKey := tmprotocrypto.PublicKey{
+		Sum: &tmprotocrypto.PublicKey_Ed25519{
+			Ed25519: consAddr.Bytes(),
+		},
+	}
+	return validator.GetOperator(), pubKey, nil
+}
+
+// ComputeValidatorUpdates computes the validator updates needed to be sent to the consumer chain to capture
+// the newly opted-in and opted-out validators, as well as validators that unbonded.
+func (k Keeper) ComputeValidatorUpdates(ctx sdk.Context,
+	currentValidators []OptedInValidator,
+	validatorAddressesToAdd []types.ProviderConsAddress,
+	validatorAddressesToRemove []types.ProviderConsAddress,
+) []abci.ValidatorUpdate {
+	var m = make(map[string]abci.ValidatorUpdate)
+
+	for _, val := range currentValidators {
+		valAddress, pubKey, err := k.getValAddressAndPublicKey(ctx, val.ProviderAddr)
+		if err != nil {
+			continue
+		}
+
+		validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, val.ProviderAddr.ToSdkConsAddr())
+		if !found {
+			continue
+		}
+
+		// if the voting power did not change since the last time the validator opted in, do not create an update
+		if val.Power == uint64(k.stakingKeeper.GetLastValidatorPower(ctx, validator.GetOperator())) {
+			continue
+		}
+
+		// if `val` has unbonded, its `GetLastValidatorPower` power returns 0
+		m[pubKey.String()] = abci.ValidatorUpdate{
+			PubKey: pubKey,
+			Power:  k.stakingKeeper.GetLastValidatorPower(ctx, valAddress),
+		}
+	}
+
+	for _, addr := range validatorAddressesToAdd {
+		valAddress, pubKey, err := k.getValAddressAndPublicKey(ctx, addr)
+		if err != nil {
+			continue
+		}
+
+		validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, addr.ToSdkConsAddr())
+		if !found {
+			continue
+		}
+
+		// if a validator is in the active set, we do not add it
+		if !validator.IsBonded() {
+			continue
+		}
+
+		m[pubKey.String()] = abci.ValidatorUpdate{
+			PubKey: pubKey,
+			Power:  k.stakingKeeper.GetLastValidatorPower(ctx, valAddress),
+		}
+	}
+
+	for _, addr := range validatorAddressesToRemove {
+		_, pubKey, err := k.getValAddressAndPublicKey(ctx, addr)
+		if err != nil {
+			continue
+		}
+
+		m[pubKey.String()] = abci.ValidatorUpdate{
+			PubKey: pubKey,
+			Power:  0,
+		}
+	}
+
+	var out []abci.ValidatorUpdate
+	for _, update := range m {
+		out = append(out, update)
+	}
+
+	// similarly to `AccumulateChanges`, we sort validators for determinism
+	sort.Slice(out, func(i, j int) bool {
+		if out[i].Power != out[j].Power {
+			return out[i].Power > out[j].Power
+		}
+		return out[i].PubKey.String() > out[j].PubKey.String()
+	})
+
+	return out
+}
+
+// ComputeNextValidators computes the next validator set that is responsible for validating on a consumer chain.
+// The returned opted-in validators correspond to the next `currentValidators`.
+func (k Keeper) ComputeNextValidators(ctx sdk.Context,
```

Comment:
Yes, that's true. I believe that splitting this into two distinct methods makes it easier to read because the `ValidateUpdate`s do not have a 1-to-1 correspodence with the next `OptedInValidator`s.
But I don't have strong feelings on this and we can combine them as well.